src/core/src/validation_util.cpp

```diff
@@ -1348,63 +1370,75 @@ bool ov::evaluate_as_partial_shape(const Output<Node>& output, PartialShape& psh
+            inputs.reserve(inputs_count);
+
+            inputs.emplace_back(element::from<label_t>(), node->get_input_shape(0));
+            std::copy(labels.begin(), labels.end(), inputs.back().data<label_t>());
+
+            for (size_t i = 1; i < inputs_count; ++i) {
+                if (node->get_input_tensor(i).has_and_set_bound()) {
+                    inputs.push_back(node->get_input_tensor(i).get_lower_value());
+                } else {
+                    return false;
+                }
             }
 
-            auto labels_constant = op::v0::Constant::create(ov::element::u64, input.get_shape(), input_labels);
-            auto idxs_htp = std::make_shared<HostTensor>(labels_constant);
-            input_tensors[i] = idxs_htp;
-        }
-    }
+            const auto& outputs_count = node->get_output_size();
+            TensorVector outputs;
+            outputs.reserve(outputs_count);
 
-    HostTensorVector output_tensors;
-    output_tensors.reserve(node->get_output_size());
-    for (size_t i = 0; i < node->get_output_size(); ++i) {
-        output_tensors.push_back(std::make_shared<HostTensor>(element::u64, node->get_output_partial_shape(i)));
-    }
+            for (size_t i = 0; i < outputs_count; ++i) {
+                const auto& partial_shape = node->get_output_partial_shape(i);
+                // Set shape for static or Shape{0} for dynamic to postpone memory allocation
+                auto shape = partial_shape.is_static() ? partial_shape.to_shape() : Shape{0};
+                outputs.emplace_back(element::from<label_t>(), shape);
+            }
 
-    if (node->evaluate(output_tensors, input_tensors)) {
-        std::transform(output_tensors.cbegin(),
-                       output_tensors.cend(),
-                       output_labels.begin(),
-                       [](const HostTensorPtr& tensor) {
-                           return std::make_shared<op::v0::Constant>(tensor)->cast_vector<label_t>();
-                       });
-        return true;
+            if (node->evaluate(outputs, inputs)) {
+                std::transform(outputs.cbegin(), outputs.cend(), output_labels.begin(), [](const Tensor& t) {
+                    // Return empty label tensor if input tensor not valid (can have Shape{0})
+                    return t ? TensorLabel(t.data<label_t>(), t.data<label_t>() + t.get_size()) : TensorLabel();
+                });
+                return true;
+            }
+        }
     }
     return false;
 }
 
-inline bool default_bound_evaluator(const Node* node, const HostTensorVector& output_values, bool is_upper) {
-    HostTensorVector input_tensors;
-    for (const auto& input : node->input_values()) {
-        if (auto bound = is_upper ? input.get_tensor().get_upper_value() : input.get_tensor().get_lower_value())
-            input_tensors.push_back(bound);
-        else
+bool ov::default_lower_bound_evaluator(const Node* node, TensorVector& output_values) {
```

Comment:
Because this flag is not required and as runtime value is checked every iteration. The decision of upper/lower bound evaluation is done during programing (compile time) but this parameter is not compile time value.
Changed to have common evaluator but without check of `is_upper`.